test/js_test.go

```diff
@@ -3596,6 +3597,134 @@ func TestJetStreamSubscribe_RateLimit(t *testing.T) {
+		{"deliver policy", nats.DeliverAll(), nats.DeliverLast()},
+		{"optional start sequence", nats.StartSequence(1), nats.StartSequence(10)},
+		{"optional start time", nats.StartTime(time.Now()), nats.StartTime(time.Now().Add(-2 * time.Hour))},
+		{"ack wait", nats.AckWait(10 * time.Second), nats.AckWait(15 * time.Second)},
+		{"max deliver", nats.MaxDeliver(3), nats.MaxDeliver(5)},
+		{"replay policy", nats.ReplayOriginal(), nats.ReplayInstant()},
+		{"max waiting", nats.PullMaxWaiting(10), nats.PullMaxWaiting(20)},
+		{"max ack pending", nats.MaxAckPending(10), nats.MaxAckPending(20)},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			durName := nuid.Next()
+			sub, err := js.PullSubscribe("foo", durName, test.first)
+			if err != nil {
+				t.Fatalf("Error on subscribe: %v", err)
+			}
+			// Once it is created, options can't be changed.
+			_, err = js.PullSubscribe("foo", durName, test.second)
+			if err == nil || !strings.Contains(err.Error(), test.name) {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+			sub.Unsubscribe()
+		})
+	}
+
+	for _, test := range []struct {
+		name string
+		cc   *nats.ConsumerConfig
+		opt  nats.SubOpt
+	}{
+		{"ack policy", &nats.ConsumerConfig{AckPolicy: nats.AckAllPolicy}, nats.AckNone()},
+		{"rate limit", &nats.ConsumerConfig{RateLimit: 10}, nats.RateLimit(100)},
+		{"flow control", &nats.ConsumerConfig{FlowControl: false}, nats.EnableFlowControl()},
+		{"heartbeat", &nats.ConsumerConfig{Heartbeat: 10 * time.Second}, nats.IdleHeartbeat(20 * time.Second)},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			durName := nuid.Next()
+
+			cc := test.cc
+			cc.Durable = durName
+			cc.DeliverSubject = nuid.Next()
+			if _, err := js.AddConsumer("TEST", cc); err != nil {
+				t.Fatalf("Error creating consumer: %v", err)
+			}
+
+			sub, err := js.SubscribeSync("foo", nats.Durable(durName), test.opt)
+			if err == nil || !strings.Contains(err.Error(), test.name) {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+			sub.Unsubscribe()
+		})
+	}
+
+	// Verify that we don't fail if user did not set it.
+	for _, test := range []struct {
+		name string
+		opt  nats.SubOpt
+	}{
+		{"description", nats.Description("a")},
+		{"deliver policy", nats.DeliverAll()},
+		{"optional start sequence", nats.StartSequence(10)},
+		{"optional start time", nats.StartTime(time.Now())},
+		{"ack wait", nats.AckWait(10 * time.Second)},
+		{"max deliver", nats.MaxDeliver(3)},
+		{"replay policy", nats.ReplayOriginal()},
+		{"max waiting", nats.PullMaxWaiting(10)},
+		{"max ack pending", nats.MaxAckPending(10)},
+	} {
+		t.Run(test.name+" not set", func(t *testing.T) {
+			durName := nuid.Next()
+			sub, err := js.PullSubscribe("foo", durName, test.opt)
+			if err != nil {
+				t.Fatalf("Error on subscribe: %v", err)
+			}
+			// If not explicitly asked by the user, we are ok
+			_, err = js.PullSubscribe("foo", durName)
+			if err != nil {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+			sub.Unsubscribe()
+		})
+	}
```

Comment:
I was curious on the transition from unset with defaults to set but with same defaults as server like:

```
# Errors as expected since server default is 512.
sub, err := js.PullSubscribe("foo", "dur2")
sub, err := js.PullSubscribe("foo", "dur2", nats.PullMaxWaiting(513))

# No error as expected.
sub, err := js.PullSubscribe("foo", "dur")
sub, err := js.PullSubscribe("foo", "dur", nats.PullMaxWaiting(512))
```

It seems it is fine which is great, added some these extra tests in this commit:
https://github.com/wallyqs/nats.go/commit/e2c6dffbb7b273eb67dc51527b246d7aae9453d0

Reply:
ok just updated